### ISSUE TITLE
[Question] Using YOLO in multiple envs in IsaacLab

### ISSUE THREAD

**Issue by shiven2001** [isaac-sim/IsaacLab#2367]
Hello, I wanted to use Yolov11 inference as an observation for training my robot about object detection. 

I am currently using a for statement which goes through every environment camera output and runs yolo on it but I am thinking it is not ideal because I am guessing it will introduce lag for the environments. the yolo inference should run in parallel for all environments instead of sequential for statement.

How to implement that?

This is my observation term custom function:

      
      import torch
      from isaaclab.envs import ManagerBasedEnv
      from isaaclab.managers import SceneEntityCfg as SceneEntityCfg
      from isaaclab.assets import Articulation
      from ultralytics import YOLO
      import os
      import numpy as np
      import cv2  # Import OpenCV
      
      yolo_path = os.path.join(
          os.environ["HOME"], "arm_isaaclab_niris/yolo_training/yolo11m.pt"
      )
      yolo_model = YOLO(yolo_path)
      
      
      def yolo_detection_highest(
          env: ManagerBasedEnv,
          asset_cfg: SceneEntityCfg = SceneEntityCfg("robot"),
          camera_name: str = "front_camera",
          target_class: str = "truck",
          conf_thres: float = 0.2,  # Add confidence threshold parameter
      ) -> torch.Tensor:
          """
          Run YOLO detection on RGB images from the specified camera and return the detection data
          for the highest-confidence object of the target class in each environment.
          """
      
          # Get the robot asset
          asset: Articulation = env.scene[asset_cfg.name]
          num_envs = asset.data.root_state_w.shape[0]
      
          # Get RGB images from the camera (shape: [num_envs, height, width, channels])
          rgb_images = env.scene[camera_name].data.output["rgb"][
              ..., :3
          ]  # [num_envs, H, W, 3]
      
          # Convert RGB images to NumPy for YOLO processing
          rgb_images_np = rgb_images.cpu().numpy().astype(np.uint8)  # [num_envs, H, W, 3]
      
          # Initialize output tensor: [num_envs, 6 params (x1, y1, x2, y2, confidence, class_id)]
          detection_tensor = torch.zeros((num_envs, 6), device=rgb_images.device)
      
          # Process each environment
          for env_idx in range(num_envs):
              # Get single environment's RGB image
              rgb_image = rgb_images_np[env_idx]  # Shape: [H, W, 3]
              """ 
              # Visualize the image if needed
              if True:
                  # Convert RGB to BGR for OpenCV (if visualizing or saving)
                  bgr_image = cv2.cvtColor(rgb_image, cv2.COLOR_RGB2BGR)
                  window_name = f"Environment {env_idx} - RGB Image"
                  cv2.imshow(window_name, bgr_image)
                  cv2.waitKey(1)  # Wait for key press to proceed
              """
              # Run YOLO inference
              results = yolo_model(rgb_image, conf=conf_thres, verbose=False)[
                  0
              ]  # Set conf threshold
              # Process YOLO results to find the highest-confidence detection
              max_confidence = 0.0
              best_detection = None
      
              for box in results.boxes:
                  # Extract class and confidence
                  class_id = int(box.cls)
                  class_name = yolo_model.names[class_id]
                  confidence = box.conf.item()
      
                  # Filter for target class
                  if class_name.lower() == target_class.lower():
                      if confidence > max_confidence:
                          # Extract bounding box coordinates
                          b = box.xyxy[0].cpu().detach().numpy()
                          x1, y1, x2, y2 = b  # Floats are fine for tensor
                          max_confidence = confidence
                          best_detection = [x1, y1, x2, y2, confidence, class_id]
      
              # Store the highest-confidence detection, if found
              if best_detection is not None:
                  detection_tensor[env_idx] = torch.from_numpy(np.array(best_detection)).to(
                      detection_tensor.device
                  )
      
          return detection_tensor

Using Isaac Sim 4.5.0

**Comment by glvov-bdai** [comment]
Hi, you should check out how the observation is computed for other pre-trained models, and match the same pattern

https://github.com/isaac-sim/IsaacLab/blob/21f7136325136ca3f6ca4e0a8125edffe5c24f7e/source/isaaclab/isaaclab/envs/mdp/observations.py#L283

https://github.com/isaac-sim/IsaacLab/blob/21f7136325136ca3f6ca4e0a8125edffe5c24f7e/source/isaaclab_tasks/isaaclab_tasks/manager_based/classic/cartpole/cartpole_camera_env_cfg.py#L165

Also, unless I am mistaken, ultralytics yolo should have a way to batch inference to run it in parallel, I'd consult the ultralytics docs for that. If you're really worried about speed you can export to an onnx with ultralytics as well; again should be well explained in their docs
